### ISSUE TITLE
Proposal: pre-compiled binaries with GitHub actions

### ISSUE THREAD

**Issue by beriberikix** [golioth/reference-design-template#53]
I'd like to propose we introduce pre-compiled binaries with the Reference Designs using GitHub Actions. #52 implements a GHA for current target boards and should be re-usable across each RD. We've had positive feedback when we introduced [pre-compiled binaries](https://blog.golioth.io/try-out-golioth-with-a-pre-compiled-binary-on-the-thingy91/) for the thingy:91 and I believe having these binaries would also help people wanting to easily test a RD. We may also consider reusing the binaries for future use-cases like HIL tests. #52 includes some in-line comments and a description of the design decisions that went into it.

I'd love feedback! Generally interested if we want to do this and any input on the implementation :pray:

**Comment by cdwilson** [comment]
I think this is a great idea! I'm all for automating as much of the release process as possible for the reference designs.

I haven't used GHA before--@beriberikix is there a way that we can test https://github.com/golioth/reference-design-template/pull/52 before it gets merged or does it need to get merged before we can try it out?

**Comment by beriberikix** [comment]
Re: testing - yes! It already ran on my PR :) https://github.com/golioth/reference-design-template/actions/workflows/build_ncs.yml

If you click on the run, you'll find the artifact zipped up there.

**Comment by szczys** [comment]
Didn't see this issue until after we started the discussion in the PR.

I'm certainly in favor of a workflow to compile [the binaries that we already make available in our releases](https://github.com/golioth/reference-design-template/releases/tag/template_v1.1.0). Those are currently manually generated, renamed, and uploaded and the potential for human error makes me nervous.

I don't think we should automatically generate binaries on PRs and Merges. PRs especially have a lot of churn to them. Merges to main should be stable, but there is no guaranteed testing in those. Our testing is done prior to a release.

If we limit the binaries to well-tested releases we put a better face on Golioth as a product and help make customer troubleshooting more manageable.

**Comment by beriberikix** [comment]
Okay, ready for another review! Notable changes:

* Switched to manual trigger. It also includes an input for the MCUboot version, defaulting to `0.0.0`
  * Unfortunately we can't test until this is merged (GH limitation.) I did test on my fork, see screenshot below.
* Files now follow a convention of `golioth-{{repo_name}}-{{board_name}}_{{file}}_{{short_git_hash}}.*`  
* Switched to uploading `merged.bin`, `app_update.bin` and kept `zephyr.elf.`

View from my fork:

![image](https://github.com/golioth/reference-design-template/assets/178376/27d0e395-8651-4c27-8089-9d945ae3ffaa)

What is not included is the notion of a release version because this doesn't yet implement a release process. This can be added in a separate PR as an additional workflow after we define our automated release process!

**Comment by szczys** [comment]
This all sounds great, nice work!

One nit from above, you mentioned `merged.bin` but the correct `merged.hex` is what's in the PR so we're all good on files.
